Phil/all natural ops catalog enhancement

psFried opened the pull request:
**Description:**

Rolls up a bunch of changes to our ops catalog, along with corresponding agent changes for handling shard failures and exposing connector statuses. Individual commit messages have all the deets, but here's a basic summary:

- Adds the currently active `build` id to the `shard` of all ops logs and stats messages
- Introduces new L1 and L2 `/events` ops collections for communicating events back to the control plane
- Reactors emit `shardFailure` events, which get materialized into the control plane
- Agent responds to `shardFailure` events by restarting failed shards with an incremental backoff (#1666)
- Connectors emit `connectorStatus` events, which gets included as part of the `/api/v1/status` response. (#1986)
- Roll up all catalog stats by every catalog name prefix, for example task `a/b/c` would have stats rolled up to `a/b/` and `a/`. (#1866)
- Adds new `txn_count` field to materialized stats, which records the number of completed transactions
- Adds `last_source_published_at` timestamp to the per-binding stats, which records the publication timestamp of the most recently processed document for that binding
- Formalizes the journal template suffix as `generation_id` and threads the value through on all inferred schemas (#1969)

Not all of this stuff is really used yet. The build id in the ops logs are intended to help with a more sophisticated status summary. And the `generation_id` will allow us to start with a blank inferred schema when re-creating a collection with the same name as an earlier one. These will be handled in future PRs.

Related connectors pr: estuary/connectors#2482

**Workflow steps:**

No changes to user-facing workflows here.

**Notes for reviewers:**

I'm still figuring out the exact steps for deploying this, and things might change here as I do. This will be a pretty complex deployment, and we'll likely want to do it during a scheduled maintenance window.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/flow/1962)
<!-- Reviewable:end -->

psFried committed:
ops: add build id to ops logs and stats

Adds the current build id to all of the logs and stats messages written by the
reactor.  The goal of this is to help the control plane identify irrelevant
logs coming from a previous task term. This makes the build id available to be
materialized.

psFried committed:
ops: introduce events derivations, connectorStatus, and shardFailures

Introduces new `/events` collections to the ops catalog, which are intended as
a general mechanism for notifying the control plane of various task-related
activities in the data plane. There is both an L1 and an L2 rollup, just like
for other ops collections.  The events collections are logically partitioned on
the task type and event type, to allow different types of handling for
different events.

An event is nothing more than a log that has a `fields.eventType` property with
a string value saying what type of event it is. Thus all events conform the
logs schema. The ops catalog itself has no opinions on what or how many event
types there are, or on the interpretation of any `fields` other than
`eventType`.

Two specific event types are also introduced here: `connectorStatus` and `shardFailure`.

`connectorStatus` events are materialized into the `connector_status` table,
keyed on the task name. The planned use case is for connectors to set the
`message` to some markdown-formatted text, and to display that as part of the
task status in the UI. If there are multiple different task shards, then only
the most recently emitted status will be shown here.  There are no reduction
annotations, so subsequent `connectorStatus` events will completely overwrite
the previous ones. This isn't for any principled reasons, but rather we just
don't have a use case for merging event fields at this time. We can always add
reductions later if we want.

The `shardFailure` events are materialized into the `shard_failures` table
using delta updates.  Inserts into this table will trigger a run of the
corresponding live specs controller, and the plan is to have controllers
respond by re-starting failed task shards. Controllers will also perform
cleanups of the `shard_failures`, though they're not able to handle that is
_all_ cases so there's also a cron job to catch any stragglers.  The main case
where controllers can't cleanup shard failures is when the failures are
inserted after the live spec (and controller) has already been deleted.

psFried committed:
reactor: emit shardFailure events

Updates the log output to set `eventType: 'shardFailure'` whenever a shard
fails. This turns it into an event, which gets materialized into the
`shard_failures` table in the control plane database.

psFried committed:
agent: controllers restart failed shards

Start handling shard failures in live specs controllers, instead of relying on
the cron-based shard kicker.

Shard failures are materialized into the `shard_failures` table, as
delta-updates.  A database trigger notifies the controller of each failure.
Controllers use an incremental backoff strategy for re-starting failed shards.
The specific implementation of the backoff should be considered a starting point.
We don't yet have the ability to trigger alerts from controllers, so we retry
indefinitely. I'm also not very confident that the backoff behavior is ideal
for tasks with many shards, so I imagine we may wish to tweak it after seeing
how it works under real-world circumstances.

The related status information was added to the `activation_status`. While the
naming of that now feels a little unfortunate, the grouping seemed to make
sense because "restarting" _is_ activation.

psFried committed:
agent: add connector_status to /status response

Pulls in the new `connector_status` into the `/api/v1/status` response. This is
represented just as an ops log, exactly as it is in the database.

psFried committed:
dekaf: add build id to ops logs

A prior commit added build ids to ops logs, so that there's a definitive record
of which version of a spec was active at the time that the logs were emitted.
This threads through that build id when emitting ops logs from dekaf.

Note that this does not detect when the build id changes in response to a
publication of the task. This is because dekaf in general has no way of
detecting this. It just fetches the latest build id whenever a connection is
established, and uses that until the connection is closed.  So changes to the
published spec won't take effect until the dekaf client re-connects, but at
least we'll now be able to tell whether that's happened by looking at the logs.

psFried committed:
ops-catalog: pre-aggregate stats by prefix

Updates the ops catalog to emit each stats document separately for each
prefix path component. For example, a stats document that's emitted for
the task `a/b/c/d` would have those same stats emitted for all of the following catalog names:

- a/
- a/b/
- a/b/c/
- a/b/c/d

The `catalog_stats` tables are altered to permit the trailing `/`, so that we
can differentiate stats for prefixes and tasks. The change to the column type
required dropping and re-creating some views. Also included are some updates to
various views and functions to account for the extra prefix stats being in the
`catalog_stats` tables.

psFried committed:
ops-catalog: add txn_count to catalog stats

Our stats documents include a count of transactions that have been completed,
but this value was not previously included in the stats summary that gets
materialized into the control plane.  This adds a `txn_count` column to the
`catalog_stats*` tables and threads the value through the ops catalog.

psFried committed:
ops-catalog: introduce `last_source_updated_at`

Introduces a new field as part of materialization and derivation stats
documents, which tracks the publication timestamp (embedded in the
`/_meta/uuid`) of the most recently read document for each binding.  This is
included in the full `taskStats` as part of the stats materialization, but they
are not materialized as any top-level columns

psFried committed:
ops-catalog: thread through collection generation id on inferred schemas

This builds on previous work, which added the id of the publication that
initially created a collection as a suffix of the journal template.
Originally, this was intended to allow correct handling of cloud storage so
that a collection (or task) can be deleted and re-created again with the same
name.  The journal template suffix will be different, so that data from the
previous generation is kept separate from data for the new generation. But a
problem remains with inferred schemas, since the inferred schemas table only
stores the collection name and not the journal template suffix.

This begins formalizing that journal template suffix as a `generation_id`, and
updates the runtime and ops catalog to thread through the generation id as part
of inferred schema updates. It is materialized as part of the
`inferred_schemas` table.

This commit stops short of actually changing how inferred schema resolution works.
Notably, the primary key of the `inferred_schemas` table is unchanged, meaning that
multiple different inferred schemas still cannot coexist for the same `collection_name`.
Instead, the `collection_generation_id` is simply materialized alongside the schema,
and we can figure later how we want that to factor into inferred schema resolutions
as part of publications.

psFried committed:
flowctl-go use max build id in test command

This had been hard-coded to a zero-valued id, which resulted in validation
errors about it being superceded. This makes it the same max value that we use
in flowctl in these scenarios.

psFried committed:
agent: update metrics endpoint to include last_source_published_at

psFried committed:
agent: update metrics endpoint to include txn_count

jgraettinger commented on go/runtime/ops_publisher.go:
```diff
@@ -90,6 +91,24 @@ func (p *OpsPublisher) PublishStats(
+// PublishConnectorLog is just like PublishLog, except it sanitizes the
+// `eventType` field of any logs, in order to prevent connectors from publishing
+// any events other than `connectorStatus`.
+func (p *OpsPublisher) PublishConnectorLog(out ops.Log) {
```
We should do this filtering / rejection closer to the connector for two reasons:

* At this point, the log has been processed by the in-process Rust runtime service and then been sent to Go.
* I intend to reduce the amount of Go runtime code as much as possible over time.

One place to filter would be `flow-connector-init`, but that's inside of the container boundary, so probably not a great idea. `crates/runtime/container.rs` has a log loop that reads structured logs from `flow-connector-init` and dispatches to `log_handler`. Probably that's the place to filter.

jgraettinger reviewed (commented):
Some initial comments. I'm starting on `agent: controllers restart failed shards` next.

jgraettinger reviewed (commented):
Okay, made it through! All pretty minor

jgraettinger commented on crates/runtime/src/derive/mod.rs:
```diff
@@ -44,14 +44,17 @@ struct Transform {
+    /// Per-transform read document stats. The u64 is the Clock (parsed from the
+    /// document uuid), which tracks the publication timestamp of the most
+    /// recently read source document for each transform.
+    read_stats: BTreeMap<u32, (DocsAndBytes, u64)>,
```
nit: I see other task types use Clock -- I'll suggest that this should to, and we newtype the raw u64 a bit earlier.

jgraettinger replied:
Yea, it just seems to be a choice of calling Clock::from_u64 "here", or calling it "there". Might want to #[inline] from_u64

jgraettinger commented on ops-catalog/catalog-stats.ts:
```diff
@@ -82,15 +88,48 @@ type StatsData = {
+function catalogPrefixes(catalogName: string): string[] {
+    let splits = catalogName.split("/");
+    let out: string[] = [];
+    for (let i = 1; i < splits.length; i++) {
```
Oh, another option would be rolling up to just `/` to represent platform-side stats, which would preserve the "`*/` is a prefix" assumption.

psFried replied:
I'm feeling a bit nervous about this kind of automatic rollup to platform-wide stats, because it might limit our ability to do shard scale-out on the stats materialization. I'm already _a bit_ nervous about the performance of the ops catalog changes here, so I'd like to push out platform-level rollups to a future PR unless we have an immediate use case for it.

psFried commented on crates/agent/src/api/public/open_metrics.rs:
```diff
@@ -361,6 +365,11 @@ define_metrics! {
         type_: COUNTER,
         help: "Total log lines indicating task failure, by task",
     },
+    TXN_COUNT= Metric {
```
It's never set for collections, except for derivations.

psFried commented on ops-catalog/catalog-stats.ts:
```diff
@@ -133,6 +133,8 @@ function mapStatsToDocsByGrain(grains: TimeGrain[], stats: StatsData): Document[
 const taskStats = (source: SourceStats): StatsData => {
     const stats: Document["statsSummary"] = {};
 
+    stats.txnCount = source.txnCount;
```
Yeah, this shouldn't be populated for collections at all, except for derivations (which I realized were missing as of the version you've reviewed). I'm not seeing a reason to try to unroll that onto collections at this point. I'm primarily just looking at `txn_count` as a way for controllers to tell whether or not a given task has been able to make any progress after a failure (for when we want to start disabling tasks automatically)

jgraettinger replied:
That's fair, I'm cool with saying only tasks have them.

psFried commented on ops-catalog/catalog-stats.ts:
```diff
@@ -82,15 +88,48 @@ type StatsData = {
+            // they contain the raw stats broken down by binding, which would
+            // otherwise result in objects with a potentially absurd number of
+            // keys after they get reduced.
+            // if (newDoc.taskStats) {
```
This is now handled by just not adding the `taskStats` in the first place, for prefix rollups. I'd meant to move this explanation up to the right spot.

psFried commented on ops-catalog/local-view.bundle.json:
```diff
@@ -3,7 +3,7 @@
       "endpoint": {
         "connector": {
-          "image": "ghcr.io/estuary/materialize-postgres:dev",
+          "image": "ghcr.io/estuary/materialize-postgres:local",
```
Yeah, this was accidental

psFried commented on supabase/migrations/20250204230547_ops_catalog_status.sql:
```diff
@@ -119,4 +119,422 @@ where
+-- advantage of the pre-aggregated catalog stats instead of aggregating them
+-- using `group by`. Also, the `today_captures` and `today_materializations`
+-- columns are removed because they were unused.
+CREATE VIEW internal.new_free_trial_tenants AS WITH hours_by_day AS (
```
Yeah, I looked a little closer at this, and this view is being queried every day, so I see no reason to look very far in the past. I updated it to have the `hours_by_day` just look at the most recent 7 days (the thought being then we could tolerate 6 failures of the cron job). And it just looks at monthly stats for the current month.

psFried committed:
agent: cleanup unused imports

psFried committed:
reactor: validate event logs from connectors

Adds validation of events that are logged by connectors, which limits them to
connectorStatus events for the currently running task. This prevents connectors
from accidentally breaking the ops catalog if they emit an event with the wrong
type or target.

psFried commented on ops-catalog/catalog-stats.ts:
```diff
@@ -82,15 +88,48 @@ type StatsData = {
+            // if (newDoc.taskStats) {
+            //     delete newDoc['taskStats'];
+            // }
+            out.push(newDoc);
```
The tests are pushed up now

psFried commented on supabase/migrations/20250204230547_ops_catalog_status.sql:
```diff
@@ -138,6 +138,14 @@ alter table public.catalog_stats_daily alter column catalog_name type text using
 alter table public.catalog_stats_monthly alter column catalog_name type text using catalog_name::text;
 alter table public.catalog_stats alter column catalog_name type text using catalog_name::text;
 
+alter table public.catalog_stats_hourly add column txn_count bigint;
```
We discussed this on VC and decided to leave `txn_count` here for now. The plan is to use it to determine whether a given task appears healthy.

psFried commented on crates/runtime/src/capture/task.rs:
```diff
@@ -132,6 +132,12 @@ impl Binding {
+        let collection_spec = collection.as_ref().context("missing collection")?;
+        // Determine the generation id of the collection, which we'll need when updating inferred schemas.
+        // For legacy collections that don't have a generation id, we'll use the zero id.
+        let collection_generation_id = tables::utils::get_collection_generation_id(collection_spec)
```
For some reason I have it in my head that `proto-flow` can't depend on `models`, though I can't remember why. But this is why I put it in `tables`, because the function returns it as a `models::Id`. It could move to `proto-flow` if we return it as a `String`. Either way seems fine to me, though I'm inclined to just leave it in `tables` for now, and just be open to moving it as we add other usages of it.

jgraettinger reviewed (approved):
LGTM
